x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/kerberos/KerberosAuthenticationToken.java

```diff
@@ -109,9 +111,31 @@ public boolean equals(final Object other) {
         return Arrays.equals(otherKerbToken.decodedToken, this.decodedToken);
     }
 
-    private static ElasticsearchSecurityException unauthorized(final String message, final Throwable cause, final Object... args) {
+    /**
+     * Creates {@link ElasticsearchSecurityException} with
+     * {@link RestStatus#UNAUTHORIZED} and cause. This also populates
+     * 'WWW-Authenticate' header with 'Negotiate' scheme.
+     * <p>
+     * If outToken is provided then that is appended to 'Negotiate ' and is used as
+     * header value for header 'WWW-Authenticate' sent to the peer in the form
+     * 'Negotiate oYH1MIHyoAMK...'. This is required by client for GSS negotiation
+     * to continue further.
+     *
+     * @param message the detail message
+     * @param cause nested exception
+     * @param outToken if non {@code null} and not empty then this will be the value
+     *            sent to the peer.
+     * @param args the arguments for the message
+     * @return instance of {@link ElasticsearchSecurityException}
+     */
+    static ElasticsearchSecurityException unauthorized(final String message, final Throwable cause, final String outToken,
```

Comment:
I'm worried that this method has too many arguments to also have a varargs `args`.
It will be easy to accidentally pass things in the wrong order. 
Maybe it should be split up so that it the _authenticate_-_negotiate_ handling takes an `ElasticsearchSecurityException` directly rather than the args to create one.